x/operator/keeper/voting_power_snapshot.go

```diff
@@ -0,0 +1,285 @@
+		value = store.Get(types.KeyForVotingPowerSnapshot(common.HexToAddress(avsAddr), ret.LastChangedHeight))
+		if value == nil {
+			return 0, nil, types.ErrNoKeyInTheStore.Wrapf("LoadVotingPowerSnapshot: fall back to the height %v", ret.LastChangedHeight)
+		}
+		k.cdc.MustUnmarshal(value, &ret)
+	}
+	return findHeight, &ret, nil
+}
+
+// RemoveVotingPowerSnapshot remove all snapshots older than the input epoch number.
+func (k *Keeper) RemoveVotingPowerSnapshot(ctx sdk.Context, avsAddr string, epochNumber int64) error {
+	if epochNumber < 0 {
+		return types.ErrParameterInvalid.Wrapf("the input epoch number is negative, epochNumber:%v", epochNumber)
+	}
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixVotingPowerSnapshot)
+	iterator := sdk.KVStorePrefixIterator(store, common.HexToAddress(avsAddr).Bytes())
+	defer iterator.Close()
+	// the retained key is used to record the snapshot that will be fallen back to
+	// by snapshots earlier than the input time.
+	var retainedKey []byte
+	for ; iterator.Valid(); iterator.Next() {
+		var snapshot types.VotingPowerSnapshot
+		k.cdc.MustUnmarshal(iterator.Value(), &snapshot)
+		_, height, _ := types.ParseVotingPowerSnapshotKey(iterator.Key())
+		if snapshot.EpochNumber > epochNumber {
+			// delete the retained key, because the snapshots that is earlier than the input time
+			// don't need to retain any old snapshot key.
+			if height == snapshot.LastChangedHeight && retainedKey != nil {
+				store.Delete(retainedKey)
+			}
+			break
+		}
+		// When height == snapshot.LastChangedHeight, it indicates that the current snapshot
+		// contains the current voting power set, so there is no need to fall back to other keys.
+		if height == snapshot.LastChangedHeight {
+			// delete the old retained key, because the key currently holding the voting power set
+			// will become the latest retained key.
+			if retainedKey != nil {
+				store.Delete(retainedKey)
+			}
+			retainedKey = iterator.Key()
+		} else {
+			store.Delete(iterator.Key())
+		}
+	}
+	return nil
+}
+
+func (k *Keeper) UpdateSnapshotHelper(ctx sdk.Context, avsAddr string, opFunc func(helper *types.SnapshotHelper) error) error {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixSnapshotHelper)
+	var snapshotHelper types.SnapshotHelper
+	value := store.Get([]byte(avsAddr))
+	if value != nil {
+		k.cdc.MustUnmarshal(value, &snapshotHelper)
+	}
+	err := opFunc(&snapshotHelper)
+	if err != nil {
+		return err
+	}
+	bz := k.cdc.MustMarshal(&snapshotHelper)
+	store.Set([]byte(avsAddr), bz)
+	return nil
+}
+
+func (k *Keeper) SetOptOutFlag(ctx sdk.Context, avsAddr string, hasOptOut bool) error {
+	opFunc := func(helper *types.SnapshotHelper) error {
+		helper.HasOptOut = hasOptOut
+		return nil // Reserve for future error handling
+	}
+	return k.UpdateSnapshotHelper(ctx, avsAddr, opFunc)
+}
+
+func (k *Keeper) SetLastChangedHeight(ctx sdk.Context, avsAddr string, lastChangeHeight int64) error {
```

Comment:
It is recommended to unify the naming style, using either **Set** or **Update** prefix for all modification functions.

Reply:
It will be renamed to `SetSnapshotHelper` in the next commit.